trieste/models/interfaces.py

```diff
@@ -241,6 +264,113 @@ def log(self) -> None:
             with tf.name_scope(f"{i}"):
                 model.log()
 
+    def reparam_sampler(self, num_samples: int) -> ReparametrizationSampler:
+        """
+        Return a reparameterization sampler providing `num_samples` samples across
+        all the models in the model stack. This is currently only implemented for
+        stacks made from models that have a :class:`BatchReparametrizationSampler`
+        as their reparameterization sampler.
+
+        :param num_samples: The desired number of samples.
+        :return: The reparametrization sampler.
+        :raise NotImplementedError: If the models in the stack do not share the
+            same :meth:`reparam_sampler`.
+        """
+
+        sampler_types = [type(model.reparam_sampler(1)) for model in self._models]
+        unique_sampler_types = set(sampler_types)
+        if len(unique_sampler_types) > 1:
+            raise NotImplementedError(
+                f"""
+                Reparameterization sampling is only currently supported for model
+                stacks built from models that use the same reparameterization sampler,
+                however, received samplers of types {unique_sampler_types}.
+                """
+            )
+        else:  # return the shared sampler rebuilt for the whole model stack
+            shared_sampler = sampler_types[0]
+            return shared_sampler(num_samples, self)
+
+
+class ReparametrizationSampler(ABC):
+    r"""
+    Thes samplers employ the *reparameterization trick* to draw samples from a
+    :class:`ProbabilisticModel`\ 's predictive distribution  across a discrete set of
+    points.
```

Comment:
we should put a reference here that explains the reparamtrization trick in greater detail